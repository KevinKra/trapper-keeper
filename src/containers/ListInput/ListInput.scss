@import "../../variables.scss";

.ListInput {
  @include container-styling(700px);
  display: flex;
  flex-direction: column;
  background-color: white;
  box-shadow: 0 0 5px $boxShadow;
  input {
    font-size: 18px;
    margin: 0.7rem 0;
    padding: 0.4rem 1rem;
    border: none;
    color: $darkFont;
    width: 100%;
    &:focus {
      outline: none;
    }
  }

  .input-title {
    width: auto;
  }

  .input-list-item {
    width: 90%;
    padding: 0 1rem 0 0;
    margin: 0;
    border: none;
    resize: none;
    overflow-y: auto;
    &:focus {
      outline: none;
    }
  }
  ::placeholder {
    font-weight: 700;
  }
  i {
    @include list-icon-styles;
    padding-left: 0.8rem;
  }
}

.list-items {
  display: flex;
  justify-content: space-between;
  align-content: center;
  vertical-align: center;
  margin: 1.2rem 0 0 0;
  border-top: 1px solid #fff;
  border-bottom: 1px solid #fff;
  i {
    align-self: center;
    color: grey;
    margin-right: 0.6rem;
    &:hover {
      cursor: pointer;
    }
  }
}

.active {
  border-top: 1px solid $bordergrey;
  border-bottom: 1px solid $bordergrey;
}

.hidden {
  visibility: hidden;
}

.btn-container {
  display: flex;
  justify-content: flex-end;
}

.btn {
  margin: 0;
  padding: 4px;
  border: 1px solid white;
  justify-content: flex-end;
  background-color: white;
  font-weight: 500;
  color: $darkFont;
  &:hover {
    cursor: pointer;
    border-radius: 5px;
    border: 1px solid rgb(240, 240, 240);
    background-color: rgb(240, 240, 240);
  }
}

.existing-note {
  display: flex;
  align-items: center;

  input {
    margin: 0;
    padding: 5px 0;
  }

  i {
    align-self: center;
    color: grey;
    margin-right: 0.6rem;
    &:hover {
      cursor: pointer;
    }
  }
  & .complete {
    text-decoration: line-through;
    color: $lightgrey;
  }
}

.complete-notes {
  border-top: 1px solid $bordergrey;
}

.modal {
  height: 100vh;
  width: 100vw;
  display: flex;
  justify-content: center;
  align-items: center;
  background-color: rgba(0, 0, 0, 0.25);
  z-index: 5;
  position: absolute;
  overflow-x: scroll;
  top: 0;
  cursor: pointer;
}
