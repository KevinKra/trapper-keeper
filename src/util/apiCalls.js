export const getAllLists = async () => {
  const res = await fetch(`${process.env.REACT_APP_BASE_URL}/api/v1/lists`);
  if (!res.ok) {
    throw new Error("Could not get lists");
  }
  return await res.json();
};

export const addNewList = async list => {
  const res = await fetch(`${process.env.REACT_APP_BASE_URL}/api/v1/lists`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(list)
  });
  if (!res.ok) {
    throw new Error("Could not save new list");
  }
  return await res.json();
};

export const getListData = async id => {
  const res = await fetch(
    `${process.env.REACT_APP_BASE_URL}/api/v1/lists/${id}`
  );
  if (!res.ok) {
    throw new Error("Could not fetch list with that id");
  }
  return await res.json();
};

export const deleteList = async id => {
  const res = await fetch(
    `${process.env.REACT_APP_BASE_URL}/api/v1/lists/${id}`,
    {
      method: "DELETE"
    }
  );
  if (!res.ok) {
    throw new Error("Could not delete card");
  }
  return await res.json();
};
