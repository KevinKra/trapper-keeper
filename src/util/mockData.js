export const mockNote = {
	id: Date.now(),
	userTask: 'Hello, world!',
	complete: false
};

export const mockList = {
	title: 'My List',
	notes: [mockNote, mockNote, mockNote]
};

export const mockLists = [mockList, mockList, mockList];
