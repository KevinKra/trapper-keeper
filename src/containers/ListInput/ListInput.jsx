import React, { Component } from 'react';
import { connect } from 'react-redux';
import { addList } from '../../actions';
import { addNewList, getListData, updateList } from '../../util/apiCalls';
import './ListInput.scss';

export class ListInput extends Component {
	state = {
		isActive: 'inactive',
		main: '',
		title: '',
		notes: []
	};

	componentDidMount() {
		this.getExistingInfo();
	}

	getExistingInfo = async () => {
		const { id } = this.props;
		if (id) {
			const list = await getListData(id);
			const { title, notes } = list;
			this.setState({ title, notes });
		}
	};

	handleFocus = () => {
		this.setState({ isActive: 'active' });
	};

	handleBlur = () => {
		this.state.displayCloseIcon ? this.setState({ isActive: 'inactive' }) : this.setState({ isActive: 'inactive' });
	};

	handleChange = e => {
		const { name, value } = e.target;
		this.setState({ [name]: value });
	};

	handleKeyPress = e => {
		if (e.key === 'Enter' && this.state.main) {
			const newNote = { id: Date.now(), userTask: this.state.main, complete: false };
			this.setState({ notes: [...this.state.notes, newNote], main: '', displayCloseIcon: false });
		}
	};

	handleSave = async () => {
		const { lists, id } = this.props;
		const { title, notes } = this.state;
		const existing = lists.find(list => list.id === parseInt(id));
		if (existing) {
			await this.updateList(existing);
			//update in redux
		} else {
			const listData = { title: title || 'Untitled List', notes };
			await this.createNewList(listData);
		}
	};

	createNewList = async listData => {
		try {
			const list = await addNewList(listData);
			this.props.addList(list);
			this.setState({ title: '', main: '', notes: [] });
		} catch (err) {
			console.log(err);
		}
	};

	updateList = async list => {
		try {
			await updateList(list);
		} catch (err) {
			console.log(err);
		}
	};

	handleToggle = id => {
		return () => {
			const { notes } = this.state;
			const note = notes.find(n => n.id === id);
			note.complete = !note.complete;
			this.setState({ notes });
		};
	};

	handleUpdate = id => {
		return e => {
			const { notes } = this.state;
			const thing = notes.find(note => note.id === id);
			thing.userTask = e.target.value;
			this.setState({ notes });
		};
	};

	render() {
		const incompleteNotes = this.state.notes.filter(n => !n.complete).map(note => {
			return (
				<div key={note.id} className="existing-note">
					<i className="material-icons" onClick={this.handleToggle(note.id)}>
						crop_square
					</i>
					<input value={note.userTask} onChange={this.handleUpdate(note.id)} />
				</div>
			);
		});

		const completeNotes = this.state.notes.filter(n => n.complete).map(note => {
			return (
				<div key={note.id} className="existing-note">
					<i className="material-icons" onClick={this.handleToggle(note.id)}>
						check
					</i>
					<input className="complete" value={note.userTask} onChange={this.handleUpdate(note.id)} />
				</div>
			);
		});

		return (
			<div className="modal">
				<form className="ListInput">
					<input
						type="text"
						name="title"
						value={this.state.title}
						autoComplete="off"
						placeholder="Title"
						className="input-title"
						onChange={this.handleChange}
					/>
					{incompleteNotes}
					{completeNotes.length > 0 && <div className="complete-notes">{completeNotes}</div>}
					<div className={`list-items ${this.state.isActive}`}>
						<i className="material-icons">crop_square</i>
						<input
							className="input-list-item"
							type="text"
							name="main"
							value={this.state.main}
							autoComplete="off"
							placeholder="List item"
							onFocus={this.handleFocus}
							onBlur={this.handleBlur}
							onChange={this.handleChange}
							onKeyPress={this.handleKeyPress}
						/>
						{this.state.main.length > 0 ? (
							<i className="material-icons">close</i>
						) : (
							<i disabled={true} className="material-icons hidden">
								close
							</i>
						)}
					</div>
					<div className="btn-container">
						<div role="button" className="btn" onClick={this.handleSave}>
							Save
						</div>
					</div>
				</form>
			</div>
		);
	}
}

export const mapStateToProps = state => ({
	lists: state.lists
});

export const mapDispatchToProps = dispatch => ({
	addList: list => dispatch(addList(list))
});

export default connect(mapStateToProps, mapDispatchToProps)(ListInput);
