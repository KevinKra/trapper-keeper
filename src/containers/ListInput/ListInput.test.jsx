import React from 'react';
import { ListInput } from './ListInput';
import { shallow } from 'enzyme';
import { getListData, addNewList, updateList } from '../../util/apiCalls';
import { mockList, mockLists, mockHistory, mockNote } from '../../util/mockData';

jest.mock('../../util/apiCalls');

describe('ListInput', () => {
	let wrapper, mockUpdateList, mockAddList;
	beforeEach(() => {
		mockUpdateList = jest.fn();
		mockAddList = jest.fn();
		wrapper = shallow(
			<ListInput
				id={mockList.id}
				lists={mockLists}
				history={mockHistory}
				updateList={mockUpdateList}
				addList={mockAddList}
			/>
		);
	});

	it('should match the snapshot', () => {
		expect(wrapper).toMatchSnapshot();
	});

	describe('componentDidMount', () => {
		it('should call getExistingInfo on call', () => {
			jest.spyOn(wrapper.instance(), 'getExistingInfo');
			wrapper.instance().componentDidMount();
			expect(wrapper.instance().getExistingInfo).toHaveBeenCalled();
		});
	});

	describe('getExistingInfo', () => {
		afterEach(() => {
			getListData.mockClear();
		});

		it.skip('should check if there is an id, and if not do nothing', async () => {
			wrapper = shallow(<ListInput id={null} />);
			const intialState = {
				isActive: false,
				main: '',
				title: '',
				notes: [],
				error: null
			};
			await wrapper.instance().getExistingInfo();
			expect(getListData).toHaveBeenCalledTimes(0);
			expect(wrapper.state()).toEqual(intialState);
		});

		it.skip('should call getListData if there is an id and set state to the mock data', async () => {
			getListData.mockImplementation(() => mockList);
			const expected = {
				isActive: false,
				main: '',
				title: mockList.title,
				notes: mockList.notes,
				error: ''
			};

			await wrapper.instance().getExistingInfo();
			expect(getListData).toHaveBeenCalledWith(mockList.id);
			expect(wrapper.state()).toEqual(expected);
		});

		it.skip('should save the error in state if it fails', async () => {
			wrapper = shallow(<ListInput id={999} />);
			getListData.mockImplementation(() => Promise.reject(new Error()));
			await wrapper.instance().getExistingInfo();
			expect(wrapper.state('error')).not.toEqual('');
		});
	});

	describe('handleChange', () => {
		it('should setState when called for the corresponding input', () => {
			let event = { target: { name: 'main', value: 'test' } };
			wrapper.find('input[name="main"]').simulate('change', event);
			expect(wrapper.state('main')).toEqual('test');

			event = { target: { name: 'title', value: 'My List' } };
			wrapper.find('input[name="title"]').simulate('change', event);
			expect(wrapper.state('title')).toEqual('My List');
		});
	});

	describe('handleKeypress', () => {
		it('should do nothing if e.key is not enter', () => {
			const event = { key: 'Not Enter' };
			wrapper.find('input[name="main"]').simulate('keypress', event);
			expect(wrapper.state('notes')).toEqual([]);
		});
		it('should do nothing if main is empty', () => {
			const event = { key: 'Enter' };
			wrapper.find('input[name="main"]').simulate('keypress', event);
			expect(wrapper.state('notes')).toEqual([]);
		});

		it('should call saveNote if both conditions are met', () => {
			jest.spyOn(wrapper.instance(), 'saveNote');
			const event = { key: 'Enter' };
			wrapper.setState({ main: 'Test content' });
			wrapper.find('input[name="main"]').simulate('keypress', event);
			expect(wrapper.instance().saveNote).toHaveBeenCalled();
		});
	});

	describe('saveNote', () => {
		it('should add a new note to state', () => {
			wrapper.setState({ main: 'Test content' });
			wrapper.instance().saveNote();
			expect(wrapper.state('notes')).toHaveLength(1);
		});
	});

	describe('handleSave', () => {
		it('should call updateList if it finds a list', async () => {
			jest.spyOn(wrapper.instance(), 'updateList');
			wrapper.setState({ title: mockList.title, notes: mockList.notes });
			await wrapper.instance().handleSave();
			expect(wrapper.instance().updateList).toHaveBeenCalledWith(mockList);
		});

		it('should call createNewList if it does not find a list', async () => {
			wrapper = shallow(<ListInput id={null} lists={mockLists} history={mockHistory} />);
			const note = { title: 'My Title', notes: [mockNote] };
			wrapper.setState({ ...note });
			jest.spyOn(wrapper.instance(), 'createNewList');
			await wrapper.instance().handleSave();
			expect(wrapper.instance().createNewList).toHaveBeenCalledWith(note);
		});
	});

	describe('createNewList', () => {
		let listData;
		beforeEach(() => {
			listData = { ...mockList };
			delete listData.id;
		});

		it('should call apiCalls.addNewList and addList if there are no errors', async () => {
			await wrapper.instance().createNewList(listData);
			expect(addNewList).toHaveBeenCalledWith(listData);
			expect(mockAddList).toHaveBeenCalled();
		});

		it('should call apiCalls.addNewList and addList if there are no errors', async () => {
			const list = await wrapper.instance().createNewList(listData);
			expect(mockAddList).toHaveBeenCalledWith(list);
		});

		it('should setState back to initial', async () => {
			const expected = {
				isActive: false,
				main: '',
				title: '',
				notes: [],
				error: ''
			};
			await wrapper.instance().createNewList(listData);
			expect(wrapper.state()).toEqual(expected);
		});

		it.skip('should set the error state if there is an error', async () => {
			getListData.mockImplementation(() => Promise.reject(new Error()));
			await wrapper.instance().createNewList(listData);
			expect(wrapper.state('error')).not.toEqual('');
		});
	});

	describe('updateList', () => {
		it('should call apiCalls.updateList on call', async () => {
			await wrapper.instance().updateList(mockList);
			expect(updateList).toHaveBeenCalledWith(mockList);
		});

		it('should call props.updateList on call', async () => {
			await wrapper.instance().updateList(mockList);
			expect(mockUpdateList).toHaveBeenCalledWith(mockList);
		});

		// ! why does this one work
		it('should set the error state if there is an error', async () => {
			updateList.mockImplementation(() => Promise.reject(new Error()));
			await wrapper.instance().updateList(mockList);
			expect(wrapper.state('error')).not.toEqual('');
		});
	});

	describe('deleteNote', () => {
		it('should remove the note with the corresponding id', () => {
			const mockNotes = [mockNote, { ...mockNote, id: 1 }, mockNote];
			wrapper.setState({ notes: mockNotes });
			wrapper.instance().deleteNote(1);
			expect(wrapper.state('notes')).toEqual(expect.not.arrayContaining([{ ...mockNote, id: 1 }]));
		});
	});
});
