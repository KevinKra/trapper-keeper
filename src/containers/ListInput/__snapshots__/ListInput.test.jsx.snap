// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListInput should match the snapshot 1`] = `ShallowWrapper {}`;
